### DISCUSSION TITLE
Some questions about the negotiation pipeline in std/col

### DISCUSSION THREAD

**Discussion by rezunli96** [yasserfarouk/scml#18] [standard-qa] [answered]
Hello, I am trying to figuring out the details of a negotiation flow in the code. Having some questions:
1. In scml2020/component/negotiation.py, in step() of class negotiator
```
   .....
        else:
            # generate buy and sell negotiations to secure inputs/outputs the step after horizon steps
            nxt = s + self._horizon + 1
            if nxt > self.awi.n_steps - 1:
                return
            self._generate_negotiations(nxt, False)
            self._generate_negotiations(nxt, True)
        if self._log:
            self.awi.logdebug_agent(f"End step:\n{pformat(self.internal_state)}")
```
Does it mean a negotiator is forced to generate negotiations in the next horizon time step?

2. Being a little bit confused about the notion of agenda in std/col negotiation. To my understanding a negotiator should propose (target_quantities(), acceptable_unit_price()) as the issues right? But why bother proposing a range of issues as supposed to be implemented in _start_negotiations()? Is a negotiation process operated normally just like in oneshot (like every agent alternatively propose() and respond())?

3. About the issue delivery time. How does the negotiators really bargaining over this issue? Is it treated as a normal issue to be negotiated like quantity and price, or is it as special one that actually is already determined (may be in step())? Also why in _trange the range is different for seller and buyer?

Thanks.

**Comment by yasserfarouk** [comment, marked as answer]
> Does it mean a negotiator is forced to generate negotiations in the next horizon time step?

The agent is no **forced** to do anything. If the agent decides to utilize one of the built-in negotiation managers then it opts to generate these negotiations. You can of course override this method and decide not to generate all or some of these negotiations.

> To my understanding a negotiator should propose (target_quantities(), acceptable_unit_price()) as the issues right? 

- Again that is not required. Agents can opt to use it. Please check this video for details:
To my understanding a negotiator should propose (target_quantities(), acceptable_unit_price()) as the issues right?  [https://youtu.be/dzGWkTL46Jk](https://youtu.be/dzGWkTL46Jk). The negotiation agenda defines the limits of quantity, price and delivery time. In oneshot these are set by the system based on current trading prices. In std/col, the agent can define them to be anything it likes. Accepting to start the negotiation implies accepting the agenda but that does not affect in any way the negotiation protocol except in limiting what is considered as valid offers. For example if the agenda specifies that quantity can be between 1 and 5 then an offer with quantity 6 will be considered invalid and will lead to immediate termination of the negotiation. 


> Is a negotiation process operated normally just like in oneshot (like every agent alternatively propose() and respond())?

- Yes it does.

> How does the negotiators really bargaining over this issue? Is it treated as a normal issue to be negotiated like quantity and price, or is it as special one that actually is already determined (may be in step())? 

Yes it is a normal negotiation issue.

> Also why in _trange the range is different for seller and buyer?

The `step` input for this method has a different semantic for buy and sell negotiations. For negotiations to buy, `step` is the time **by which** the agent needs the products and for negotiations to sell, it is the first step at which it can sell these products. That explains the difference you see in the code.

By the way, which track are you interested in?

**Reply by rezunli96** [reply]
Ahh okay I see. Making delivery date an issue makes the whole game much more interesting and challenging. Wondering how we can design a smart negotiation strategy for that (like why would one agent want to execute an order after several days).
My plan is to start from the oneshot track, which I to my understand is closer to the classical automated negotiation scenarios, while still being envisioning the other two tracks.
Anyway, thanks for your response 👍
